deepchecks/base/dataset.py

```diff
@@ -297,6 +297,9 @@ def check_compatible_labels(self):
 
     # Validations:
 
+    # TODO: error messages is totally unrelated to the Dataclass itself
+    # I think we need to refactor this methods
```

Comment:
You mean having the validations external to the Dataset? Possibly, they relate to the Dataset but not all datasets need to pass the same validations

Reply:
I mean that if I were the user of the library and if I called the `validate_index` method (outside of the check context) and got a message like this - `Check {check_name} requires dataset to have an index column`, I would have been confused. How check instances are related to the fact that I want to make sure that my dataset has an index?

the second thing that bothers me is that it does not help with type hints. Calling `validate_index` does not help the static type checker to infer that accessing `self.index_col()` will actually return `pd.Series` and not `Optional[pd.Series]`

Reply:
I get it, first point makes sense. @ItayGabbay WDYT? We can open a task for that next sprint

About type hints - is there a way to do in a way that helps? Note that type hints are more for us, they don't really help the user as far as I know

Reply:
> About type hints - is there a way to do in a way that helps? 

will try to create PR for that

> Note that type hints are more for us, they don't really help the user as far as I know

1. for users it means: better autocomplete, good documentation
2. I would say that they are not for us but for the static type checker 

benefits of extensively using type annotations (copied from mypy docs)
```
1.Static typing can make programs easier to understand and maintain. Type declarations can serve as machine-checked documentation. This is important as code is typically read much more often than modified, and this is especially important for large and complex programs.
2.Static typing can help you find bugs earlier and with less testing and debugging. Especially in large and complex projects this can be a major time-saver.
3.Static typing can help you find difficult-to-find bugs before your code goes into production. This can improve reliability and reduce the number of security issues.
```